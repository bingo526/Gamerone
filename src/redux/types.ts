import { RouterState } from 'connected-react-router';
import { AuthState } from './auth/types';
import { ProfileState } from './profile/types';
import { SettingsState } from './settings/types';
import { PostState } from './post/types';
import { PostFormState } from './post-form/types';
import { DialogState } from './dialogs/types';
import { RequestStatusState } from './request-status/types';
import { MiscState } from './misc/types';

export const INIT_STATE = 'root/INIT_STATE';
export const LOAD_STATE = 'root/LOAD_STATE';

export const LOAD_INITIAL_GAMES = 'root/LOAD_INITIAL_GAMES';
export const LOAD_NEXT_GAMES = 'root/LOAD_GAMES';

export const LOAD_EXPERIENCES = 'root/LOAD_EXPERIENCES';
export const LOAD_EXPERIENCE_SUMMARY = 'root/LOAD_EXPERIENCE_SUMMARY';

export interface InitStateAction {
  type: typeof INIT_STATE;
}

export type RootStateActions = InitStateAction;

export interface RootState {
  Auth: AuthState;
  Profile: ProfileState;
  Settings: SettingsState;
  Post: PostState;
  PostForm: PostFormState;
  Dialog: DialogState;
  RequestStatus: RequestStatusState;
  router: RouterState;
  Misc: MiscState;
}
