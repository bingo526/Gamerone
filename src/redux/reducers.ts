import Auth from './auth/reducer';
import Profile from './profile/reducer';
import Settings from './settings/reducer';
import Post from './post/reducer';
import RequestStatus from './request-status/reducer';
import PostForm from './post-form/reducer';
import Dialog from './dialogs/reducer';
import Misc from './misc/reducer';

export default {
  Auth,
  Profile,
  Settings,
  Post,
  PostForm,
  Dialog,
  RequestStatus,
  Misc,
};
