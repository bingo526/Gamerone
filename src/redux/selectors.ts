/**
 * Here comes the combined selectors from different states
 */

import { createSelector } from 'reselect';

import { selectCurrentUserId } from './auth/selectors';
import sort from 'fast-sort';

import {
  selectCurrentProfileUser,
  selectCurrentProfileNetworks,
  selectCurrentProfileSponsors,
  selectCurrentProfileProducts,
  selectCurrentProfileLayout,
  selectProfileGames,
  selectProfileGears,
  selectProfileNowPlaying,
  selectResolvedContent,
  selectResolvedContentId,
  selectProfileLatestAchievements,
  selectProfileGearsResponse,
  selectProfileExperience,
  selectProfileExperienceSummary,
} from './profile/selectors';

import {
  selectSettingsSponsors,
  selectSettingsProducts,
  selectSettingsProfileUser,
  selectProfileSocialNetworks,
  selectLayoutSettings,
  selectSettingsGames,
  selectSettingsGears,
  selectSettingsNowPlaying,
  selectSettingsGearsResponse,
  selectSettingsExperience,
  selectSettingsExperienceSummary,
} from './settings/selectors';

import { selectStatus } from './request-status/selectors';

import { DEFAULT_USER_VISIBILITY } from 'utils/constants';
import { NowPlayingModel } from 'models/NowPlayingModel';
import { selectFeed, selectFeedResponse } from './post/selectors';
import { FEED_DATA, LOAD_FEED_REQUEST } from './post/types';
import { PostModel } from 'models/post';
import { Achievement, Post } from 'interfaces';

/**
 * These selectors below are used to select states effectively based on the currently viewing profile.
 *
 * - If the user is on their OWN profile, they select data from `Settings` state
 * - Otherwise, they select data from `Profile` state
 *
 * Note: We initially started by having separate pages like Settings and Profile (hopefully they make sense what they were for), and now we incorporated Settings and Profile features into one page (although still WIP), but the separated logic behind will remain unchanged (or we will consider to update later)
 */

/**
 * determines whether the currently viewing profile is my profile or not
 */
export const selectIsSelfProfile = createSelector(
  [selectResolvedContentId, selectCurrentUserId],
  (resolvedId, authUserId) => resolvedId === authUserId,
);

export const selectProfileUser = createSelector(
  [selectCurrentProfileUser, selectSettingsProfileUser],
  (profileUser, settingsUser) =>
    settingsUser.id === profileUser?.id ? settingsUser : profileUser,
);

/**
 * resolves social networks to be shown on Profile page
 */
export const selectSocialNetworks = createSelector(
  [
    selectIsSelfProfile,
    selectProfileSocialNetworks,
    selectCurrentProfileNetworks,
  ],
  (isSelfProfile, settingsNetworks, profileNetworks) =>
    isSelfProfile ? settingsNetworks : profileNetworks,
);

/**
 * resolves sponsors to be shown on Profile page
 */
export const selectSponsors = createSelector(
  [selectIsSelfProfile, selectSettingsSponsors, selectCurrentProfileSponsors],
  (isSelfProfile, settingsSponsors, profileSponsors) =>
    isSelfProfile ? settingsSponsors : profileSponsors,
);

/**
 * resolves products to be shown on Profile page
 */
export const selectProducts = createSelector(
  [selectIsSelfProfile, selectSettingsProducts, selectCurrentProfileProducts],
  (isSelfProfile, settingsProducts, profileProducts) =>
    isSelfProfile ? settingsProducts : profileProducts,
);

/**
 * resolves profile layout to be shown on Profile page
 */
export const selectProfileLayout = createSelector(
  [selectIsSelfProfile, selectLayoutSettings, selectCurrentProfileLayout],
  (isSelfProfile, settingsLayout, profileLayout) => {
    const layout = isSelfProfile ? settingsLayout : profileLayout;
    if (layout && !layout.visibility)
      layout.visibility = DEFAULT_USER_VISIBILITY;
    return layout;
  },
);

/**
 * resolves games to be shown on Profile page
 */
export const selectGames = createSelector(
  [selectIsSelfProfile, selectSettingsGames, selectProfileGames],
  (isSelfProfile, settingsGames, profileGames) =>
    isSelfProfile ? settingsGames : profileGames,
);

export const selectGamesStatus = createSelector(
  [selectIsSelfProfile, selectStatus],
  (isSelf, status) =>
    status.get(`${isSelf ? 'settings' : 'profile'}/LOAD_PAGE_REQUEST/games`),
);

export const selectGamesResponse = createSelector(
  [
    selectIsSelfProfile,
    selectSettingsGearsResponse,
    selectProfileGearsResponse,
  ],
  (isOwner, settingsResponse, profileResponse) =>
    isOwner ? settingsResponse : profileResponse,
);

export const selectHasMoreGames = createSelector(
  selectGamesResponse,
  (response) => (response ? response.currentPage >= response.lastPage : false),
);

/**
 * resolves experiences to be shown on Experience page
 */
export const selectExperience = createSelector(
  [selectIsSelfProfile, selectSettingsExperience, selectProfileExperience],
  (isSelfProfile, settingsExperience, profileExperience) =>
    isSelfProfile ? settingsExperience : profileExperience,
);

export const selectExperienceStatus = createSelector(
  [selectIsSelfProfile, selectStatus],
  (isSelf, status) =>
    status.get(`${isSelf ? 'settings' : 'profile'}/GET_EXPERIENCES_REQUEST`),
);

export const selectExperienceSummary = createSelector(
  [
    selectIsSelfProfile,
    selectSettingsExperienceSummary,
    selectProfileExperienceSummary,
  ],
  (isSelfProfile, settingsExperienceSummary, profileExperienceSummary) =>
    isSelfProfile ? settingsExperienceSummary : profileExperienceSummary,
);

export const selectExperienceSummaryStatus = createSelector(
  [selectIsSelfProfile, selectStatus],
  (isSelf, status) =>
    status.get(
      `${
        isSelf
          ? 'settings/GET_USER_ATTRIBUTES_REQUEST'
          : 'profile/GET_EXPERIENCE_SUMMARY_REQUEST'
      }`,
    ),
);

/**
 * resolves gears to be shown on Profile page
 */
export const selectGears = createSelector(
  [selectIsSelfProfile, selectSettingsGears, selectProfileGears],
  (isSelfProfile, settingsGears, profileGears) =>
    isSelfProfile ? settingsGears : profileGears,
);

export const selectGearsStatus = createSelector(
  [selectIsSelfProfile, selectStatus],
  (isSelf, status) =>
    status.get(`${isSelf ? 'settings' : 'profile'}/LOAD_PAGE_REQUEST/gears`),
);

/**
 * resolves now playing data to be shown on Profile page
 */
export const selectNowPlaying = createSelector(
  [selectIsSelfProfile, selectSettingsNowPlaying, selectProfileNowPlaying],
  (isSelfProfile, settingsNowPlaying, profileNowPlaying) =>
    (isSelfProfile ? settingsNowPlaying : profileNowPlaying) ||
    new NowPlayingModel(),
);

/**
 * resolves now playing data to be shown on Profile page
 */
export const selectLatestAchievements = createSelector(
  [selectProfileLatestAchievements],
  (profileLatestAchievements) => {
    if (profileLatestAchievements) {
      return sort(profileLatestAchievements).desc((a: Achievement) => a.date);
    }
    return [];
  },
);

/**
 * Select resolved content type
 *
 * private | user | club | game
 */
export const selectResolvedContentType = createSelector(
  [selectIsSelfProfile, selectResolvedContent],
  (isSelf, content) =>
    (isSelf ? 'private' : content?.contentType || 'private') as FEED_DATA,
);

/**
 * Select currently viewing profile feed whether it's user, game, or club
 */
export const selectProfileFeed = createSelector(
  [selectFeed, selectResolvedContentType, selectResolvedContentId],
  (feedData, contentType, contentId) => {
    if (contentId in feedData[contentType]) {
      return sort([...feedData[contentType][contentId]]).desc(
        (p: Post) => p.createdAt,
      );
    }
    return contentType === 'private'
      ? sort([...feedData.private[0]]).desc((p: Post) => p.createdAt)
      : [];
  },
);

/**
 * Return the rest of the feed excluding the latest post
 */
export const selectProfileFeedRest = createSelector(selectProfileFeed, (feed) =>
  feed.slice(1),
);

export const selectHasMoreProfileFeed = createSelector(
  [selectFeedResponse, selectResolvedContentType, selectResolvedContentId],
  (feedResponse, contentType, contentId) => {
    if (contentId in feedResponse[contentType]) {
      return feedResponse[contentType][contentId]?.offset !== 0;
    }
    return contentType === 'private'
      ? feedResponse.private[0]?.offset !== 0
      : false;
  },
);

export const selectProfileFeedStatus = createSelector(
  [selectStatus, selectResolvedContentType],
  (status, contentType) => status.get(LOAD_FEED_REQUEST + '/' + contentType),
);

export const selectProfileLatestPost = createSelector(
  selectProfileFeed,
  (feed) => (feed.length > 0 ? (feed[0] as PostModel) : new PostModel()),
);
