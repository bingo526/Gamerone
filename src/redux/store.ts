import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { createBrowserHistory as createHistory } from 'history';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga';
import reducers from './reducers';
import rootSaga from './sagas';

const history = createHistory();
const sagaMiddleware = createSagaMiddleware();
const routeMiddleware = routerMiddleware(history);
const middlewares = [thunk, sagaMiddleware, routeMiddleware];

const composeEnhancers =
  (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) || compose;

const store = createStore(
  combineReducers({
    ...reducers,
    router: connectRouter(history),
  }),
  composeEnhancers(applyMiddleware(...middlewares)),
);
sagaMiddleware.run(rootSaga);
export { store, history };
