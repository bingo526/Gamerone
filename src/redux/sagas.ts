import { selectCurrentUserId } from './auth/selectors';
import {
  all,
  select,
  take,
  race,
  takeLatest,
  put,
  fork,
  call,
} from 'redux-saga/effects';

import authSagas from './auth/saga';
import profileSagas from './profile/saga';
import settingsSagas from './settings/saga';
import postSagas from './post/saga';
import postFormSagas from './post-form/saga';
import miscSagas from './misc/saga';

import {
  LOAD_INITIAL_GAMES,
  LOAD_NEXT_GAMES,
  LOAD_EXPERIENCES,
  LOAD_EXPERIENCE_SUMMARY,
} from './types';
import { selectCurrentProfileId } from './profile/selectors';
import { selectIsSelfProfile } from './selectors';
import {
  CHECK_AUTHORIZATION_SUCCESS,
  CHECK_AUTHORIZATION_ERROR,
} from './auth/types';

import SettingsActions from './settings/actions';
import ProfileActions from './profile/actions';
import { getToken } from 'utils/token';

export function* waitUntilAuthFinishes() {
  const currentUserId = yield select(selectCurrentUserId);
  const token = getToken().get('idToken');
  if (token && currentUserId === 0) {
    yield race({
      success: take(CHECK_AUTHORIZATION_SUCCESS),
      error: take(CHECK_AUTHORIZATION_ERROR),
    });
  }
}

function* loadUserExperienceSummary() {
  yield takeLatest(LOAD_EXPERIENCE_SUMMARY, function* () {
    yield call(waitUntilAuthFinishes);

    const userId = yield select(selectCurrentProfileId);
    const isOwner = yield select(selectIsSelfProfile);

    if (isOwner) {
      yield put(SettingsActions.getUserAttributes());
    } else if (userId) {
      yield put(ProfileActions.getExperienceSummary(userId));
    }
  });
}

function* loadUserExperiences() {
  yield takeLatest(LOAD_EXPERIENCES, function* () {
    yield call(waitUntilAuthFinishes);

    const userId = yield select(selectCurrentProfileId);
    const isOwner = yield select(selectIsSelfProfile);

    if (isOwner) {
      yield put(SettingsActions.getExperiences());
    } else if (userId) {
      yield put(ProfileActions.getExperiences(userId));
    }
  });
}

function* loadInitialGames() {
  yield takeLatest(LOAD_INITIAL_GAMES, function* () {
    yield call(waitUntilAuthFinishes);

    const userId = yield select(selectCurrentProfileId);
    const isOwner = yield select(selectIsSelfProfile);

    if (isOwner) {
      yield put(SettingsActions.loadInitialPage('games'));
    } else if (userId) {
      yield put(ProfileActions.loadInitialPage('games'));
    }
  });
}

function* loadNextGames() {
  yield takeLatest(LOAD_NEXT_GAMES, function* () {
    const userId = yield select(selectCurrentProfileId);
    const isOwner = yield select(selectIsSelfProfile);

    if (isOwner) {
      yield put(SettingsActions.loadNextPage('games'));
    } else if (userId) {
      yield put(ProfileActions.loadNextPage('games'));
    }
  });
}

export default function* rootSaga() {
  yield all([
    fork(loadInitialGames),
    fork(loadNextGames),

    fork(loadUserExperiences),
    fork(loadUserExperienceSummary),

    authSagas(),
    profileSagas(),
    settingsSagas(),
    postSagas(),
    postFormSagas(),
    miscSagas(),
  ]);
}
