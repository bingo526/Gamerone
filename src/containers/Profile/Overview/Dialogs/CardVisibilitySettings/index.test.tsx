import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { store } from 'redux/store';
import CardVisibilitySettingsModal from './index';

// Snapshot
test('cards visibility section initial', () => {
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <CardVisibilitySettingsModal />
      </MemoryRouter>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// DOM testing
afterEach(cleanup);
it('check and save', () => {
  const { queryByLabelText, getByText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <CardVisibilitySettingsModal />
      </MemoryRouter>
    </Provider>,
  );

  expect(queryByLabelText(/sponsors/i)).toBeFalsy();
  expect(queryByLabelText(/gear/i)).toBeFalsy();

  act(() => {
    fireEvent.click(getByText('Save'));
  });

  expect(queryByLabelText(/gear/i)).toBeTruthy();
  expect(queryByLabelText(/sponsors/i)).toBeTruthy();
});

/**
 * Mock input event to do check
 *
 * @return {Object}
 */
const renderCardVisibilityForm = () => {
  return render(
    <Provider store={store}>
      <MemoryRouter>
        <CardVisibilitySettingsModal />
      </MemoryRouter>
    </Provider>,
  );
};

/**
 * Fill in and submit the form
 *
 * @param {Object} utils
 * @param {String} sponsors
 * @param {String} socialNetworks
 * @param {String} gear
 * @param {String} store
 * @param {String} nowPlaying
 * @param {String} achievements
 */
const fillAndSubmitForm = (
  utils: import('@testing-library/react').RenderResult<
    typeof import('@testing-library/dom/queries')
  >,
  {
    sponsors = true,
    socialNetworks = false,
    gear = true,
    store = true,
    nowPlaying = false,
    achievements = true,
  }: {
    sponsors: boolean;
    socialNetworks: boolean;
    currentlyPlaying: boolean;
    achievements: boolean;
    gear: boolean;
    store: boolean;
  },
) => {
  const { getByRole, getByLabelText } = utils;

  act(() => {
    fireEvent.change(getByLabelText(/sponsors/i), {
      target: { value: sponsors },
    });
    fireEvent.change(getByLabelText(/socialNetworks/i), {
      target: { value: socialNetworks },
    });
    fireEvent.change(getByLabelText(/nowPlaying/i), {
      target: { value: nowPlaying },
    });
    fireEvent.change(getByLabelText(/achievements/i), {
      target: { value: achievements },
    });
    fireEvent.change(getByLabelText(/gear/i), {
      target: { value: gear },
    });
    fireEvent.change(getByLabelText(/store/i), {
      target: { value: store },
    });
    fireEvent.submit(getByRole('button', { name: 'Save' }));
  });
};

test('update layoutsettings', async () => {
  const utils = renderCardVisibilityForm();
  const visibility = {
    sponsors: true,
    socialNetworks: false,
    nowPlaying: true,
    achievements: true,
    gear: true,
    store: true,
  };

  // mock out window.fetch for the test
  const spy = jest.spyOn(window, 'fetch').mockImplementation((url) => {
    let success = {};
    if (url.endsWith('profile/layout-settings')) {
      success = visibility;
    }
    return Promise.resolve({
      status: 200,
      json: () => success,
    });
  });

  act(() => {
    fireEvent.click(utils.getByText('Save'));
  });

  fillAndSubmitForm(utils, visibility);

  spy.mockRestore();
});
