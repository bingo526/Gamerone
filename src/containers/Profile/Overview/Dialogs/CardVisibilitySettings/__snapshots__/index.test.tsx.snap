// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards visibility section initial 1`] = `" "`;
